src/common/fetchClient.ts

```diff
@@ -0,0 +1,154 @@
+export interface FetchClientConfig {
+  baseURL?: string;
+  headers?: Record<string, string>;
+}
+
+export interface FetchRequestConfig<D = any> {
+  params?: Record<string, string>;
+  headers?: Record<string, string>;
+  body?: D;
+}
+
+export interface FetchResponse<T = any> extends Response {
+  data: T;
+}
+
+export interface FetchResponseErrorData<T = any> {
+  status: number;
+  headers: Record<string, string>;
+  data: T;
+}
+
+export class FetchResponseError extends Error {
+  readonly response: FetchResponseErrorData;
+
+  constructor(response: FetchResponseErrorData) {
+    super();
+    this.response = response;
+  }
+}
+
+const defaultConfig: FetchClientConfig = {
+  baseURL: "",
+  headers: {},
+};
+
+export default class FetchClient {
+  config: FetchClientConfig;
+
+  constructor(config?: FetchClientConfig) {
+    this.config = {
+      ...defaultConfig,
+      ...config,
+    };
+  }
+
+  async request(
+    method: string,
+    path: string,
+    config: FetchRequestConfig = {},
+  ): Promise<FetchResponse> {
+    const url = this.buildUrl(path, config.params);
+    const headers = {
+      ...this.config.headers,
+      ...(config.headers ?? {}),
+    };
+
+    const response = await fetch(url, {
+      method,
+      headers,
+      body: config.body
+        ? this.prepareRequestBody(config.body, headers)
+        : undefined,
+    });
+    const data = await this.getResponseData(response);
+
+    if (!response.ok) {
+      throw new FetchResponseError({
+        status: response.status,
+        data,
+        headers: this.headersToObject(response.headers),
+      });
+    }
+
+    return {
+      ...response,
+      data,
+    };
+  }
+
+  async get(path: string, config: FetchRequestConfig): Promise<FetchResponse> {
+    return this.request("GET", path, config);
+  }
+
+  async post(path: string, config: FetchRequestConfig): Promise<FetchResponse> {
+    return this.request("POST", path, config);
+  }
+
+  async put(path: string, config: FetchRequestConfig): Promise<FetchResponse> {
+    return this.request("PUT", path, config);
+  }
+
+  async delete(
+    path: string,
+    config: FetchRequestConfig,
+  ): Promise<FetchResponse> {
+    return this.request("DELETE", path, config);
+  }
+
+  private buildUrl(path: string, params?: FetchRequestConfig["params"]): URL {
+    const url = new URL(this.config.baseURL + path);
+
+    if (params) {
+      Object.entries(params).forEach(([key, value]) =>
+        url.searchParams.append(key, value),
+      );
+    }
+
+    return url;
+  }
+
+  private prepareRequestBody(
+    data: any,
+    headers: Record<string, string>,
+  ): string | FormData {
+    if (headers?.["Content-Type"]?.includes("application/json")) {
+      if (typeof data === "string") {
+        return data;
+      }
```

Comment:
Can I double check what scenario this nested conditional here is intended to handle? Asking because I was getting a bit confused following the conditionals flow here. For example, if the content type was set to "text/plain", and the data passed in was a string, we would still JSON.stringify() and wasn't sure if that's the expected behavior. 

One other thing that jumped out was here we check the content type header specifically for "application/json", but not for "multipart/form-data" when handling `FormData` below?

I was wondering if we can simplify the logic here, and do the thing based on the data passed in? Something like:
```
if (!data) return undefined;
if (typeof data === "string") return data;
if (data instance of FormData) return data;
return JSON.stringify(data);
```

I may be missing some nuances here, so let me know what you think!

Reply:
Whoops, must've messed up the logic while I was refactoring it a bit. The goal was to check if the data was already stringified, but I think the logic you suggested is a lot simpler and understandable. I'll use that instead